src/libraries/System.IO/tests/Stream/Stream.CopyToSpanTests.cs

```diff
@@ -0,0 +1,528 @@
+        [InlineData(42)]
+        [InlineData(100000)]
+        public void CopyToAsync_StreamToken_ExpectedBufferSizePropagated(int length)
+        {
+            using var src = new CustomMemoryStream_BufferSize();
+            src.Write(new byte[length], 0, length);
+            src.Position = 0;
+
+            Assert.Equal(length, ((Task<int>)src.CopyToAsync((_, __, ___) => default, null, length, default(CancellationToken))).Result);
+        }
+
+        private sealed class CustomMemoryStream_BufferSize : MemoryStream
+        {
+            public override Task CopyToAsync(Stream destination, int bufferSize, CancellationToken cancellationToken) =>
+                Task.FromResult(bufferSize);
+        }
+
+        [Theory]
+        [MemberData(nameof(CopyTo_TestData))]
+        public void CopyTo_AllDataCopied(MemoryStream input)
+        {
+            using var src = input;
+            src.Write(Enumerable.Range(0, 10000).Select(i => (byte)i).ToArray(), 0, 256);
+            src.Position = 0;
+
+            using var dst = new MemoryStream();
+            src.CopyTo((span, _) => dst.Write(span), null, 4096);
+
+            Assert.Equal<byte>(src.ToArray(), dst.ToArray());
+        }
+
+        [Theory]
+        [MemberData(nameof(CopyTo_TestData))]
+        public async Task CopyToAsync_AllDataCopied(MemoryStream input)
+        {
+            using var src = input;
+            src.Write(Enumerable.Range(0, 10000).Select(i => (byte)i).ToArray(), 0, 256);
+            src.Position = 0;
+
+            using var dst = new MemoryStream();
+            await src.CopyToAsync((memory, _, ___) => dst.WriteAsync(memory), null, 4096, default);
+
+            Assert.Equal<byte>(src.ToArray(), dst.ToArray());
+        }
+
+        private sealed class CustomMemoryStream : MemoryStream
+        {
+            private readonly bool _spanCopy;
+            private readonly bool _sync;
+
+            public CustomMemoryStream(bool spanCopy, bool sync)
+            : base()
+            {
+                _spanCopy = spanCopy;
+                _sync = sync;
+            }
+
+            public override void CopyTo(Stream destination, int bufferSize)
+            {
+                if (_sync)
+                {
+                    CopyToInternal(destination, bufferSize);
+                }
+                else
+                {
+                    CopyToAsyncInternal(destination, bufferSize, CancellationToken.None).GetAwaiter().GetResult();
+                }
+            }
+
+            public override Task CopyToAsync(Stream destination, int bufferSize, CancellationToken cancellationToken)
+            {
+                if (!_sync)
+                {
+                    return CopyToAsyncInternal(destination, bufferSize, cancellationToken);
+                }
+                else
+                {
+                    try
+                    {
+                        cancellationToken.ThrowIfCancellationRequested();
```

Comment:
nit: this is just test code, so it doesn't really matter, but generally we want such exceptions to come out in the returned task rather than thrown synchronously, e.g.
```
if (cancellationToken.IsCancellationRequested)
{
    return Task.FromCanceled(cancellationToken);
}
```